File: 438.找到字符串中所有字母异位词.cpp
/*
 * @lc app=leetcode.cn id=438 lang=cpp
 *
 * [438] 找到字符串中所有字母异位词
 */

// @lc code=start
#include<vector>
#include<string>
using namespace std;
class Solution {
public:

    vector<int> feature(const string& s, int _start = 0, int _end = -1){
        if(_end == -1) _end = s.size();
        vector<int> ret(26, 0);
        for(int i=_start; i<_end; ++i) ++ret[s[i] - 'a'];
        return ret;
    }

    vector<int> findAnagrams(string s, string p) {
        int lenp = p.size(), tmp = s.size() - lenp + 1;
        if(s.size() < lenp) return {};
        auto feature_p = feature(p);
        auto feature_s = feature(s, 0, lenp);
        vector<int> ret;
        if(feature_s == feature_p) ret.push_back(0);
        for(int i=1; i<tmp; ++i){
            --feature_s[s[i-1] - 'a'];
            ++feature_s[s[i+lenp-1] - 'a'];
            if(feature_s == feature_p) ret.push_back(i);
        }
        return ret;
    }
};
// @lc code=end

File: 438.找到字符串中所有字母异位词.py
#
# @lc app=leetcode.cn id=438 lang=python3
#
# [438] 找到字符串中所有字母异位词
#

# @lc code=start
from typing import List, Optional

class Solution:

    @staticmethod
    def feature(s: str, _start: int = 0, _end: Optional[int] = None) -> List[int]:
        start, end = _start, len(s) if _end is None else _end
        ret = [0] * 26
        offset = ord('a')
        for i in range(start, end):
            ret[ord(s[i]) - offset] += 1
        return ret

    def findAnagrams(self, s: str, p: str) -> List[int]:
        lenp = len(p)
        if len(s) < lenp: return []
        feature_p = self.feature(p)
        feature_s = self.feature(s, 0, lenp)
        ret: List[int] = []
        if feature_s == feature_p: ret.append(0)
        offset = ord('a')
        for i in range(1, len(s) - lenp + 1):
            feature_s[ord(s[i-1]) - offset] -= 1
            feature_s[ord(s[i+lenp-1]) - offset] += 1
            if feature_s == feature_p: ret.append(i)
        return ret

# @lc code=end

print(Solution().findAnagrams(s = "cbaebabacd", p = "abc"))
print(Solution().findAnagrams(s = "abab", p = "ab"))
print(Solution().findAnagrams("aaaaaaaaaa", "aaaaaaaaaaaaa"))